### ISSUE TITLE
Fitting Compound Models containing multiplied units fails

### ISSUE THREAD

**Issue by kecnry** [astropy/astropy#12320]
### Description

Creating a Compound model that multiplies the underlying output units returns expected units when evaluating, but fails during fitting because its trying to coerce all units to the data units and not accounting for the operators between the internal sub-models.

### Expected behavior

Fitting should succeed, with consistent behavior between units from calling evaluate and accepted during fitting.

### Actual behavior

Calling evaluate returns the expected units, but those same units as inputs while fitting raises errors.

### Steps to Reproduce

Below is a simplified (although non-physical) example:

```
import astropy.units as u
from astropy.modeling import models
from astropy.modeling.fitting import LevMarLSQFitter
import numpy as np

fitter = LevMarLSQFitter()

x = np.linspace(0,1,101) * u.s
y = np.linspace(5,10,101) * u.m * u.kg / u.s

m1 = models.Linear1D(slope=5*u.m/u.s/u.s, intercept=1.0*u.m/u.s
m2 = models.Linear1D(slope=0.0*u.kg/u.s, intercept=10.0*u.kg) 
c = m1 * m2
```

`c(x)` gives the expected units (m * kg / s), but `fitter(c, x, y)` fails.  Under-the-hood, the following is being called:

```
c._parameter_units_for_data_units({'x': x.unit}, {'y': y.unit})
```

which is evaluating to:

```
{'intercept_0': Unit("kg m / s"),
 'slope_0': Unit("kg m / s2"),
 'intercept_1': Unit("kg m / s"),
 'slope_1': Unit("kg m / s2")}
 ```
 
 In other words, `c._parameter_units_for_data_units` is attempting to make sure that each sub-model results in the passed output units, whereas it should make sure that the units _after applying operators_ result in the passed output units.  The following (in this case - obviously the general case is much more complicated) would make sense to be consistent with the forward model:
 
```
m1._parameter_units_for_data_units({'x': x.unit}, {'y': y.unit / m2(x).unit})
m2._parameter_units_for_data_units({'x': x.unit}, {'y': y.unit / m1(x).unit})
```

### System Details

macOS-10.15.7-x86_64-i386-64bit
Python 3.9.6 (default, Aug 18 2021, 12:38:10) 
[Clang 10.0.0 ]
Numpy 1.20.3
pyerfa 2.0.0
astropy 5.0.dev1030+g09720bb32
Scipy 1.7.1
Matplotlib 3.4.3

**Comment by pllim** [comment]
cc @WilliamJamieson @nden @perrygreenfield
